File: hdk/go/plugin/request.go
package plugin

import (
	"strings"

	"github.com/bhakha-services/naagmani-plugins/hdk/go/protocol"
)

// Message is a single chat message.
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// Request represents an AI inference request.
type Request struct {
	Model    string    `json:"model"`
	Messages []Message `json:"messages"`
	Stream   bool      `json:"stream"`
}

// PromptText returns the combined text of all user messages in the request.
func (r *Request) PromptText() string {
	var parts []string
	for _, m := range r.Messages {
		if m.Role == "user" {
			parts = append(parts, m.Content)
		}
	}
	return strings.Join(parts, "\n")
}

// AddMessage appends a new message to the chat history.
func (r *Request) AddMessage(role, content string) {
	r.Messages = append(r.Messages, Message{
		Role:    role,
		Content: content,
	})
}

func requestFromProtocol(req *protocol.HookRequest) *Request {
	if req == nil {
		return nil
	}
	r := &Request{
		Model:  req.Model,
		Stream: req.Stream,
	}
	for _, m := range req.Messages {
		r.Messages = append(r.Messages, Message{
			Role:    m.Role,
			Content: m.Content,
		})
	}
	return r
}

func (r *Request) toProtocol() *protocol.HookRequest {
	if r == nil {
		return nil
	}
	proto := &protocol.HookRequest{
		Model:  r.Model,
		Stream: r.Stream,
	}
	for _, m := range r.Messages {
		proto.Messages = append(proto.Messages, protocol.HookMessage{
			Role:    m.Role,
			Content: m.Content,
		})
	}
	return proto
}

File: hdk/node/src/plugin/request.ts
import { HookMessage, HookRequest } from "../protocol/messages";

export interface Message {
  role: string;
  content: string;
}

export class Request {
  model: string;
  messages: Message[];
  stream: boolean;

  constructor(protoReq?: HookRequest) {
    this.model = protoReq?.model || "";
    this.messages = (protoReq?.messages || []).map((m) => ({
      role: m.role,
      content: m.content,
    }));
    this.stream = Boolean(protoReq?.stream);
  }

  /**
   * Combined text of all user messages in the request.
   */
  promptText(): string {
    return this.messages
      .filter((m) => m.role === "user")
      .map((m) => m.content)
      .join("\n");
  }

  /**
   * Append a new message to the chat history.
   */
  addMessage(role: string, content: string): void {
    this.messages.push({ role, content });
  }

  toProtocol(): HookRequest {
    return {
      model: this.model,
      messages: this.messages.map((m) => ({ role: m.role, content: m.content })),
      stream: this.stream,
    };
  }
}
